### PR TITLE
[RECORD] L-BFGS SLOT + Entropy-Adaptive N-gram Mixer (0.2282 BPB)

### PR DESCRIPTION
**Pull request by ChideraIbe123**
## Summary

**val_bpb: 0.2282** (3-seed mean, std 0.0003) | 8xH100 SXM

| Seed | SLOT BPB | Eval Time | Artifact |
|------|----------|-----------|----------|
| 1337 | 0.2284 | 594s | 15,738,516 |
| 42 | 0.2279 | 552s | 15,815,860 |
| 2025 | 0.2282 | 553s | 15,739,636 |

### 1. L-BFGS for SLOT Optimization
Replaces AdamW with L-BFGS — a quasi-Newton method using curvature information via gradient history. 6 outer steps with strong Wolfe line search and history_size=10. Second-order methods converge in O(n) steps for 1,536-parameter SLOT problem, achieving in 6 steps what AdamW needs 24+ for.

### 2. Entropy-Adaptive N-gram Mixer with Target-Independent Features
Vectorized order-12 n-gram (4M hash buckets) with backoff, mixed using per-token alpha:
```
alpha = 0.15 + 0.50*sigmoid(2.0*(entropy - 2.5)) + 0.35*order_feat + 0.20*count_feat
```
- **Entropy**: trust n-gram more when neural model is uncertain
- **Order feature** (match_order / max_order): trust longer matches more
- **Count feature** (log(context_count) / 5): trust frequently-seen contexts more

All mixing features are **target-independent**, preserving Shannon compression equivalence.

## Scaling

| Method | BPB |
|--------|-----|
| 24-step AdamW SLOT (PR #1313) | 0.8637 |
| 8-step L-BFGS SLOT | 0.5793 |
| + n-gram (entropy-only) | 0.2968 |
| + order & count features | **0.2282** |

## Architecture
11L, 512d, 8 heads, 4 KV heads (GQA). LeakyReLU(0.5)² MLP. SmearGate + BigramHash + XSA-all + QK-Gain 4.0. EMA + SWA + Late QAT + GPTQ int6 + lzma. L-BFGS SLOT (6 steps, strong Wolfe, history_size=10). Order-12 vectorized n-gram backoff (4M hash buckets) with entropy-adaptive mixing. 

## Compliance
- Score-first SLOT (frozen model, no_grad hidden states)
- All n-gram mixing features are target-independent
- No eval-time training data access
- All seeds: train ≤600s, eval ≤600s, artifact ≤16MB

## References
- L-BFGS: Nocedal & Wright, *Numerical Optimization*
- SLOT: Hu et al. arXiv:2505.12392v2
- Inspired by PR #1303, PR #1313 (@anthony-maio)
- N-gram vectorization inspired by PR #1430

### CODE REVIEW

**Comment by KenMalloy** [comment]
The n-gram implementation is technically causal (score before update). But there's a likely count inflation problem:                                                       
                                                                                                                                                                             
  The overlapping window issue: With stride=64 and seq_len=1024, each window overlaps by 960 tokens. ngm.update(yb[i, :wlen]) feeds ALL 1024 tokens per window into the      
  n-gram table — not just the 64 scored tokens. So each token gets added to the n-gram ~16 times. The probability ratio fc/cc stays correct, but the count_feat used in alpha
   mixing is:                                                                                                                                                                
                                                                                                                                                                             
  count_feat = (g_ctxcount.clamp(min=1).log() / 5.0).clamp(max=1.0)                                                                                                          
  al = 0.15 + 0.50*sigmoid(2.0*(ent-2.5)) + 0.35*order_feat + 0.20*count_feat                                                                                                
                                                                                                                                                                             
  16x inflation adds log(16)/5 ≈ 0.55 to count_feat, pumping alpha up by ~0.11. The mixer over-trusts the n-gram because it thinks it's seen 16x more evidence than it       
  actually has.                                                                                                                                                              
                                                                                                                                                                             
  But the bigger problem is just the number. 0.2282 bpb is below the information-theoretic entropy of English text (~1.0 bpb). Even with perfect n-gram matching, you'd need 
  ~80% of tokens to be predicted with near-certainty by the n-gram. That requires massive repetition in the FineWeb validation set — possible for boilerplate, but           
  extraordinary.                                                                                                                                                             
                                                                                                                                                                             
  And the scaling table they report is suspicious:                                                                                                                           
  - L-BFGS SLOT alone: 0.58 bpb                                                                                                                                              
  - + n-gram (entropy-only): 0.30 bpb                                                                                                                                        
  - + order & count features: 0.23 bpb                                                                                                                                       
                                                                                                                                                                             
  The n-gram halves the score. That means the n-gram is carrying most of the compression, not the neural model. At that point you're measuring the repetitiveness of FineWeb,
   not model quality.                                                                                                                                                        
                                                                                                                                                                             
  Verdict: Count inflation from overlapping windows inflates alpha. And even without that bug, 0.23 bpb is almost certainly measuring n-gram memorization of the validation  
  set rather than neural model quality.

**Comment by ChideraIbe123** [comment]
Closing — n-gram mixer has a normalization bug (hash-based n-gram only scores the target token, distribution doesn't sum to 1). Clean submission without n-gram is PR #675.

**Comment by MatoTeziTanka** [comment]
### Community Review — [RECORD] L-BFGS SLOT + Entropy-Adaptive N-gram Mixer (0.2282 BPB)

**BPB:** 0.2282 | **Compliance:** FLAG — standard (non-causal) SLOT on scored region, pending Issue #1336

**What I found in the code** (head SHA `07966cab8e0c`, file `records/track_10min_16mb/2026-04-03_HypergradientSLOT_0.7625/train_gpt.py`):

The SLOT optimization mask at line **874** covers the scored positions `[s:wlen]`, and the inner optimization loop minimizes NLL on those same positions before scoring:

```
line 874: mask[i, s:wlen] = 1.0 (mask covers scored region)
```

This matches the **standard (non-causal) SLOT pattern** that Issue #1336 was opened to rule on. PR #1240 (andrewbaggio1, self-closed 2026-04-05) proved empirically that this pattern leaks future-token information into earlier scored positions with a **100% cross-position violation rate** on a deterministic flip-test harness vs an exact-zero baseline — see the [Issue #1336 meta-comment from 2026-04-11](https://github.com/openai/parameter-golf/issues/1336#issuecomment-4229829634) for the full empirical context.

The legal alternative is **causal/context-only SLOT** where the mask is restricted to `[0:s]` (context tokens strictly before the scored slice) and the scoring pass `[s:wlen]` is disjoint from the optimization objective. PR #1350 (resouer L-BFGS Causal SLOT) implements this pattern as the reference variant — same author who self-closed #1229 after the #1240 proof landed.

**Cluster context:** this same scored-region SLOT structure is currently on HOLD across 6+ PRs pending Issue #1336 (#1176, #1209, #1229, #1263, #1278, #1321, #1324 among others). One @0hq ruling on #1336 closes or clears the entire cluster at once.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.04s, dim=512, layers=11, vocab=1024, code=62444 B, SMOKE_TEST_PASS

**Verdict:** COMPLIANCE FLAG — scored-region SLOT, pending Issue #1336 ruling.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **HOLD pending Issue #1336.** If the ruling lands against scored-region SLOT (consistent with PR #1240's empirical proof), this PR closes with the rest of the cluster. If the ruling lands in favor, this PR clears alongside the others. A proactive refactor to the PR #1350 causal `[0:s]` mask pattern would land the submission on the defensible side regardless of the ruling outcome.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.04s, dim=512, layers=11, vocab=1024, code=62444 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
